### ISSUE TITLE
Incorrect NUnit1011 when TestCaseSource references a base class member

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit.analyzers#320]
Repros in 0.4.0–0.6.0, not present in 0.3.0

```cs
using System.Collections.Generic;
using NUnit.Framework;

class Base
{
    public static IEnumerable<string> TargetFrameworks { get; } = new[] { "net48", "netcoreapp3.1" };
}

class Derived : Base
{
    // ❌ NUnit1011 The TestCaseSource argument 'TargetFrameworks' does not specify an existing member
    //              ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓
    [TestCaseSource(nameof(TargetFrameworks))]
    public void Test(string targetFramework)
    {
    }
}
```

**Comment by jnm2** [comment]
The analyzer should follow NUnit's requirements rather than C# visibility rules. This might matter if the base class member is internal, protected, private, etc.

**Comment by manfred-brands** [comment]
@jnm2 I added detection in base class, as long as the member is not declared private. Let me know if you think NUnit's visibility is different. I couldn't find it at [TestCaseSource](https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcasesource.html)

**Comment by jnm2** [comment]
@manfred-brands We should follow the behavior of https://github.com/nunit/nunit/blob/3a47c23b9fb85cdabac6ef73179751496e659a64/src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs#L247. IIRC, FlattenHierarchy doesn't bring private members from the base class even though NonPublic is specified, but I can't remember what it does with e.g. protected members. Would you like me to check?

**Comment by jnm2** [comment]
I also don't mind it being a linter-like warning along the lines of: even though this works it's best practice to model test case sources as public conceptually to make them available to a hypothetical caller. In other words, more restrictive on visibility than NUnit is. But we might still prefer to show the error rather than the warning when we know for sure that there will be a runtime error.

**Comment by manfred-brands** [comment]
@jnm2 

> We should follow the behavior of https://github.com/nunit/nunit/blob/3a47c23b9fb85cdabac6ef73179751496e659a64/src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs#L247.

This code explicitly checks for _instance_ members. Despite that the documentation says it *must* be static. Later on it throws when the returned member is non-static.
`FlattenHierarchy` will return `public` and `protected` members.

My PR searching up the hierarchy ignoring `private` members matches this.

**Comment by jnm2** [comment]
@manfred-brands Very cool, thanks for checking that! I was ignoring the part about instance members for the sake of the question. 

I tested whether NUnit let the test pass when the base test case source had these accessibility modifiers, just to be sure that I understood the reflection code I was seeing. This was the result:
✔ `public`
✔ `internal`
✔ `protected`
✔ `protected internal`
✔ `private protected`
❌ `private`

Looks like this is what your PR does too. :+1:
